### PR TITLE
Fix wrapper subclass serialization with custom sizes / strides

### LINKED ISSUE #130154: Wrapper subclasses with python dispatch size / stride impls cannot be serialized
Repro:

```python
import torch
import torch.utils._pytree as pytree
from torch.utils._python_dispatch import return_and_correct_aliasing


# A simple tensor subclass that holds two tensors internally, and runs every op on both tensors.
# It also sets dispatch_sizes_strides_policy="sizes" to get python dispatch for size info.
class TwoTensor2(torch.Tensor):
    @staticmethod
    def __new__(cls, a, b):
        assert (
            a.device == b.device
            and a.layout == b.layout
            and a.requires_grad == b.requires_grad
            and a.dtype == b.dtype
        )
        # I guess it would be more accurate to represent the shape as torch.cat(a, b).shape
        shape = a.shape
        kwargs = {}
        kwargs["strides"] = a.stride()
        kwargs["storage_offset"] = a.storage_offset()
        kwargs["device"] = a.device
        kwargs["layout"] = a.layout
        kwargs["requires_grad"] = a.requires_grad
        kwargs["dtype"] = a.dtype
        # NB: this policy setting triggers the bug
        kwargs["dispatch_sizes_strides_policy"] = "sizes"
        out = torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)

        assert a.shape == b.shape
        assert a.stride() == b.stride()
        assert a.storage_offset() == b.storage_offset()
        return out

    def __init__(self, a, b):
        self.a = a
        self.b = b

    def __repr__(self):
        a_repr = repr(self.a)
        b_repr = repr(self.b)
        return f"TwoTensor2({a_repr}, {b_repr})"

    def __tensor_flatten__(self):
        return ["a", "b"], None

    @staticmethod
    def __tensor_unflatten__(inner_tensors, meta, outer_size, outer_stride):
        assert meta is None
        a, b = inner_tensors["a"], inner_tensors["b"]
        return TwoTensor2(a, b)

    @classmethod
    def __torch_dispatch__(cls, func, types, args, kwargs):
        if kwargs is None:
            kwargs = {}
        args_a = pytree.tree_map_only(TwoTensor2, lambda x: x.a, args)
        args_b = pytree.tree_map_only(TwoTensor2, lambda x: x.b, args)

        kwargs_a = pytree.tree_map_only(TwoTensor2, lambda x: x.a, kwargs)
        kwargs_b = pytree.tree_map_only(TwoTensor2, lambda x: x.b, kwargs)

        out_a = func(*args_a, **kwargs_a)
        out_b = func(*args_b, **kwargs_b)
        assert type(out_a) == type(out_b)
        out_a_flat, spec = pytree.tree_flatten(out_a)
        out_b_flat = pytree.tree_leaves(out_b)
        # for aten ops that return non-tensors, just assume that
        # our two inner tensors return the same value
        out_flat = [
            TwoTensor2(o_a, o_b) if isinstance(o_a, torch.Tensor) else o_a
            for o_a, o_b in zip(out_a_flat, out_b_flat)
        ]
        out = pytree.tree_unflatten(out_flat, spec)
        return return_and_correct_aliasing(func, args, kwargs, out)


t = TwoTensor2(torch.randn(3), torch.randn(3))
# accessing this caches a PyCapsule on the object 
t.shape
# TypeError: cannot pickle 'PyCapsule' object
torch.save(t, "tensor.pt")
```

This happens because of custom size PyCapsule caching on the object:
https://github.com/pytorch/pytorch/blob/e7ab7b83bc36825c102987a26b4a199f97674561/torch/csrc/PyInterpreter.cpp#L603-L725

cc @mruberry @mikaylagawarecki @ezyang @albanD

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137126
* #137125
* #137124
* #137099
* #137031
* __->__ #137030

Fixes #130154

This PR takes the strategy outlined in the above issue and clears out any cached sizes / strides PyCapsules before serialization. This affects the default subclass serialization logic.

The PyCapsule issue also affects `deepcopy`, so that's fixed here as well.

Note: I originally tried utilizing a context manager to remove / restore cached PyCapsules after serialization, but in practice the state returned from `_reduce_ex_internal()` references the actual `tensor.__dict__()`, so the problem persists once the cached values are restored. Instead, we have to be careful to remove the cached values in the right place so they're not re-cached when pulling out size / stride information for serialization.

### CODE REVIEW

File: torch/_tensor.py [hunk 1]
```diff
@@ -227,6 +241,9 @@ def __reduce_ex__(self, proto):
         if has_torch_function_unary(self):
             return handle_torch_function(Tensor.__reduce_ex__, (self,), self, proto)
         func, args = self._reduce_ex_internal(proto)
+        # sizes / strides cache needs to be cleared here because it'll just be recached
+        # if cleared earlier. Note that state references the -actual- tensor dict.
+        _clear_cached_sizes_strides(self)
```
Comment (albanD) [inline torch/_tensor.py:246]: Should we make this a method on the class that user can also override? Or we expect them to override `__reduce_ex__` to do cleanup before they call `super().__reduce_ex__(` ?

File: torch/testing/_internal/common_subclass.py [hunk 2]
```diff
@@ -196,6 +254,16 @@ def __init__(self, name, create_fn, closed_under_ops=True):
         self.closed_under_ops = closed_under_ops
 
 
+# Helper function to create a subclass of the given class and possibly cache sizes / strides.
+def _create_and_access_shape(cls, shape):
+    sub = cls(torch.randn(shape))
+    # NB: Wrapper subclasses with custom dispatched sizes / strides cache this info
+    # on the first call via non-serializable PyCapsules.
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply torch/testing/_internal/common_subclass.py:261]: will do, it's also for deepcopy tests :)

File: torch/_tensor.py (hunk 1 above)
Comment (jbschlosser) [reply torch/_tensor.py:246]: Hm I like the idea of making this an overrideable class method. Is the name good or should it be generalized some to cover the general concept of clearing non-serializable cached stuff, wdyt?

File: torch/_tensor.py [hunk 3]
```diff
@@ -78,6 +78,22 @@ def _rebuild_from_type_v2(func, new_type, args, state):
 # torch/_C/__init__.pyi.in to add a type annotation for your method;
 # otherwise, it will not show up in autocomplete.
 class Tensor(torch._C.TensorBase):
+    def _clear_non_serializable_data(self):
```
Comment (albanD) [inline torch/_tensor.py:81]: Please add a docstring.
In particular mentioning that a subclass that overrides this  MUST call the super() variant

Comment (albanD) [inline torch/_tensor.py:81]: Given that this is for content that is valid to clear, I think `cached` should be in the name

**Review by mikaylagawarecki — DISMISSED** [review]

**Review by albanD — APPROVED** [review]
SGTM !

**Pull request closed without merging**